convert2rhel/applock.py

```diff
@@ -0,0 +1,182 @@
+    process is still running, it raises ApplicationLockedError. If the
+    process is not running, it removes the file and tries to lock it
+    again.
+
+    For safety, unexpected conditions, like garbage in the file or
+    bad permissions, are treated as if the application is locked,
+    because something is obviously wrong.
+    """
+
+    def __init__(self, name):
+        # Our application name
+        self._name = name
+        # Do we think we locked the pid file?
+        self._locked = False
+        # Our process ID
+        self._pid = os.getpid()
+        # Path to the file that contains the process id
+        self._pidfile = os.path.join(_DEFAULT_LOCK_DIR, self._name + ".pid")
+
+    def __str__(self):
+        if self._locked:
+            status = "locked"
+        else:
+            status = "unlocked"
+        return "%s PID %d %s" % (self._pidfile, self._pid, status)
+
+    def _try_create(self):
+        """Try to create the lock file. If this succeeds, the lock file
+        exists and we created it. If an exception other than the one
+        we expect is raised, re-raises it.
+
+        :returns: True if we created the lock, False if we didn't.
+        """
+        #
+        # Create a temporary file that contains our PID and attempt
+        # to link it to the real pidfile location. The link will fail if
+        # the file already exists; this avoids a race condition when
+        # two processes attempt to create the file simultaneously. This
+        # also guarantees that the lock file contains valid data.
+        #
+        # Note that NamedTemporaryFile will clean up the file it created,
+        # but the lockfile we created by doing the link will stay around.
+        #
+        with tempfile.NamedTemporaryFile(mode="w", suffix=".pid", prefix=self._name, dir=_DEFAULT_LOCK_DIR) as f:
+            f.write(str(self._pid) + "\n")
+            f.flush()
+            try:
+                os.link(f.name, self._pidfile)
+            # In Python 3 this could be changed to catch FileExistsError.
+            except OSError as exc:
+                if exc.errno == errno.EEXIST:
+                    return False
+                raise exc
+        loggerinst.debug("%s." % self)
+        return True
+
+    @property
+    def is_locked(self):
+        """Test whether this object is locked."""
+        return self._locked
+
+    @staticmethod
+    def _pid_exists(pid):
+        """Test whether a particular process exists."""
+        try:
+            # Bulletproofing: avoid killing init or all processes.
+            if pid > 1:
+                os.kill(pid, 0)
+        except OSError as exc:
+            #
+            # The only other (theoretical) possibility is EPERM, which
+            # would mean the process exists.
+            #
+            if exc.errno == errno.ESRCH:
+                return False
+        return True
+
+    def try_to_lock(self, recursive=False):
```

Comment:
```suggestion
    def try_to_lock(self, _recursive=False):
```

By convention, private variables, parameters, functions, etc, start with an underscore.  Since this is just used internally to guard against calling ourselves more than once, rename the parameter.

Also note, this is a place in python where I'd use iteration instead of recursion.  It seems more natural when dealing with something that has a determinable number of iterations (and recursion is natural when the number of iterations cannot be determined).

Reply:
Name changed.

I disagree somewhat with the idea of changing this to user iteration. I can see the argument back when we were trying to lock 5 times, but the text description of what we're doing is "Try to lock the file; if the flie exists and is invalid, remove it and try again." To my mind the recursive call adheres to that description more precisely.

Reply:
Iteration and recursion are equivalents.  So I don't think that description matches recursion better than iteration.  (I might even say that iteration is closer as the difference between the two runs is built into the syntax of the loop whereas the difference between the two runs in recursion is in the explicit parameter that you pass to the function to stop the recursion.)  This could also be partially due to python's performance characteristics favoring iteration over recursion (function calls are relatively expensive in python compared to other languages.  It doesn't have built in tail recursion optimizations, for instance).

In general,  internal-only parameters are a code smell that tells you that there's probably something that you should change with your design (in python).

(Don't change this now, though.  I'm okay with this going in as is for MVP).